Sources/SwiftDocC/Model/Rendering/Variants/VariantPatchOperation.swift

```diff
@@ -56,26 +56,87 @@ public enum VariantPatchOperation<Value: Codable> {
-                result.append(contentsOf: newValue)
-            case .remove:
-                result.removeAll()
+// The synthesized implementation is sufficient for this conformance.
+extension VariantPatchOperation: Equatable where Value: Equatable {}
+
+// MARK: Applying patches
+
+/// A type that can be transformed by incrementally applying variant patch operations.
+protocol VariantCollectionPatchable {
+    /// Apply an "add" patch operation to the value
+    mutating func add(_ other: Self)
+    /// Apply a "remove" patch operation to the value
+    mutating func remove()
+}
+
+extension Optional: VariantCollectionPatchable where Wrapped: VariantCollectionPatchable {
+    mutating func add(_ other: Wrapped?) {
+        guard var wrapped, let other else { return }
+        wrapped.add(other)
+        self = wrapped
+    }
+    
+    mutating func remove() {
+        self = nil
+    }
+}
+
+extension Array: VariantCollectionPatchable {
+    mutating func add(_ other: [Element]) {
+        append(contentsOf: other)
+    }
+    
+    mutating func remove() {
+        self.removeAll()
+    }
+}
+
+extension String: VariantCollectionPatchable {
+    mutating func add(_ other: String) {
+        append(contentsOf: other)
+    }
+    
+    mutating func remove() {
+        self.removeAll()
+    }
+}
+
+extension VariantCollection where Value: VariantCollectionPatchable {
+    /// Returns the transformed value after applying the patch operations for all variants that match the given source language to the default value.
+    /// - Parameters:
+    ///   - language: The source language that determine what variant's patches to apply to the default value.
+    /// - Returns: The transformed value, or the default value if no variants match the given source language.
+    func value(for language: SourceLanguage) -> Value {
+        applied(to: defaultValue, for: [.interfaceLanguage(language.id)])
+    } 
+    
+    /// Returns the transformed value after applying the patch operations for all variants that match the given traits to the default value.
+    /// - Parameters:
+    ///   - traits: The traits that determine what variant's patches to apply to the default value.
+    /// - Returns: The transformed value, or the default value if no variants match the given traits.
+    func value(for traits: [RenderNode.Variant.Trait]) -> Value {
+        applied(to: defaultValue, for: traits)
+    }
+    
+    /// Returns the transformed value after applying the patch operations for all variants that match the given traits to the original value.
+    /// - Parameters:
+    ///   - originalValue: The original value to transform.
+    ///   - traits: The traits that determine what variant's patches to apply to the original value.
+    /// - Returns: The transformed value, or the original value if no variants match the given traits.
+    func applied(to originalValue: Value, for traits: [RenderNode.Variant.Trait]) -> Value {
```

Comment:
I changed this abstraction because the previous one wasn't all that useful. 

There are several properties in the navigator builder that need to determine what the value is for a given trait, some for the render node:

```swift
var topicSections: [TaskGroupRenderSection] {
    wrapped.topicSectionsVariants.value(for: traits)
}
var defaultImplementationsSections: [TaskGroupRenderSection] {
    wrapped.defaultImplementationsSectionsVariants.value(for: traits)
}
```
but most for the render node metadata:
```swift
var title: String? {
    wrapped.titleVariants.value(for: traits)
}
var navigatorTitle: [DeclarationRenderSection.Token]? {
    wrapped.navigatorTitleVariants.value(for: traits)
}
var fragments: [DeclarationRenderSection.Token]? {
    wrapped.fragmentsVariants.value(for: traits)
}
var externalID: String? {
    wrapped.externalIDVariants.value(for: traits)
}
var roleHeading: String? {
    wrapped.roleHeadingVariants.value(for: traits)
}
var symbolKind: String? {
    wrapped.symbolKindVariants.value(for: traits)
}
var platforms: [AvailabilityRenderItem]? {
    wrapped.platformsVariants.value(for: traits)
}
```

The previous abstraction operated on the individual `VariantCollection.Variant` (as opposed to operating on the full `VariantCollection`). This API wasn't particularly useful for two reasons:
- it was the callers responsibility to iterate over the variants for a certain trait
- the API was only available when the value was a (range replaceable) collection

The first drawback meant that to determine the Objective-C value of a render node, each caller needs to access the default value, filter and iterate over the variants, and incrementally apply the patch:

```swift
var topicSections: [TaskGroupRenderSection] {
    var patchedValue = wrapped.topicSectionsVariants.defaultValue
    for variant in wrapped.topicSectionsVariants.variants where variant.traits == traits {
        patchedValue = variant.applyingPatchTo(patchedValue)
    }
    return patchedValue
}
```

Most of this is boilerplate compared to the new API

```swift
var topicSections: [TaskGroupRenderSection] {
    wrapped.topicSectionsVariants.value(for: traits)
}
```

The second trackback meant that to determine the Objective-C value of render node's metadata's platforms, the caller _couldn't_ use the current API because the `Value` of `VariantCollection<[AvailabilityRenderItem]?>` is an optional value and not a collection. Thus, the caller not only needs to access the default value, filter and iterate over the variants, and incrementally apply the patch but also needs to switch over there patch and incrementally apply it depending on the type of `Value`:

```swift
var platforms: [AvailabilityRenderItem]? {
    var patchedValue = wrapped.platformsVariants.defaultValue
    for variant in wrapped.platformsVariants.variants where variant.traits == traits {
        for patch in variant.patch {
            switch patch {
            case .replace(let value):
                patchedValue = value
            case .add(nil):
                break // Adding nil does nothing
            case .add(let value?):
                if patchedValue == nil {
                    patchedValue = value
                } else {
                    patchedValue!.append(contentOf: value)
                }
            case .remove:
                patchedValue = nil
            }
        }
    }
    return patchedValue
}
```

The two steps to address all this boilerplate is:
- Create an interface for types that a patch can be applied to 
- Move the API to `VariantCollection` (instead of `VariantCollection.Value`)

The new API did both these things, thus obsoleting the old API. 
Further, most of the time the caller doesn't want to apply the patch to an _arbitrary_ value. It makes the most sense to apply the patch to the default value for that variant collection. Because of this, I wrapped the `applied(to:for:)` API with `value(for:)`.

Reply:
That makes sense, thanks for the detailed explanation. The new API is definitely easier to read.